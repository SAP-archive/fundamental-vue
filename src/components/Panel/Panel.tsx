import './Panel.css';
import { Slot, Component, DefaultSlot, Prop, Base } from '@/core';

interface Props {
  title?: string | null;
  description?: string | null;
  span?: number | null;
  condensed?: boolean;
  condensedFooter?: boolean;
}

@Component('Panel')
@DefaultSlot('Panel Body')
@Slot('actions', 'Panel Actions')
@Slot('filters', 'Custom Panel Filters')
@Slot('footer', 'Custom Panel Footer')
export class Panel extends Base<Props> {
  @Prop({ type: String, default: null })
  public title!: string | null;

  @Prop({ type: String, default: null })
  public description!: string | null;

  @Prop({ type: Number, default: null })
  public span!: number | null;

  @Prop('whether the panel body is condensed (has no padding)', { type: Boolean, default: false })
  public condensed!: boolean;

  @Prop('whether the panel footer is condensed (has no padding)', { type: Boolean, default: false })
  public condensedFooter!: boolean;

  public render() {
    const body = this.$slots.default;
    const title = this.title;
    const description = this.description;
    const actions = this.$slots.actions;
    const hasActions = !!actions;
    const needsHeader = title != null || description != null || actions != null;
    const filters = this.$slots.filters;
    const footer = this.$slots.footer;
    const hasFooter = !!footer;
    const hasFilters = !!filters;

    function renderHeader() {
      const needsHead = title != null || description != null;
      return (
        <div class='fd-panel__header'>
          {needsHead &&
            <div class='fd-panel__head'>
              {title && <h1 class='fd-panel__title'>{title}</h1>}
              {description && <p class='fd-panel__description'>{description}</p>}
            </div>
          }
          {hasActions && <div class='fd-panel__actions'>{actions}</div>}
        </div>
      );
    }

    const bodyClasses = () => {
      return {
        'fd-panel__body': true,
        'vf-panel__condensed': this.condensed,
      };
    };

    const footerClasses = () => {
      return {
        'fd-panel__footer': true,
        'vf-panel__condensed': this.condensedFooter,
      };
    };

    return (
      <div class={this.classObject}>
        {needsHeader && renderHeader()}
        {hasFilters && <div class='fd-panel__filters'>{filters}</div>}
        <div class={bodyClasses()}>{body}</div>
        {hasFooter && <div class={footerClasses()}>{footer}</div>}
      </div>
    );
  }

  get classObject() {
    const spanObject = () => {
      const span = this.span;
      if (typeof span === 'number') {
        return { [`fd-has-grid-column-span-${span}`]: true };
      }
      return {};
    };
    return {
      ...spanObject(),
      'fd-panel': true,
    };
  }
}
