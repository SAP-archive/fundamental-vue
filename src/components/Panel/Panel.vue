<template>
  <div :class="classes">
    <!-- HEADER -->
    <div v-if="needsHeader" class="fd-panel__header">
      <div v-if="needsHead" class="fd-panel__head">
        <h1 v-if="hasTitle" class="fd-panel__title">
          <slot name="title">
            {{ title }}
          </slot>
        </h1>
        <p v-if="hasDescription" class="fd-panel__description">
          <slot name="description">
            {{ description }}
          </slot>
        </p>
      </div>
      <div v-if="hasActions" class="fd-panel__actions">
        <slot name="actions" />
      </div>
    </div>
    <div v-if="$slots.filters != null" class="fd-panel__filters">
      <slot name="filters" />
    </div>
    <div v-if="$slots.default" :class="bodyClasses"><slot /></div>
    <div v-if="$slots.footer" :class="footerClasses">
      <slot name="footer" />
    </div>
  </div>
</template>

<script>
import { paddingClasses } from "./../../directives/design-system-utilities";

const isValidColSpan = value => value >= 2 && value <= 6;

const PANEL_CLASS = "fd-panel";

export default {
  name: "FdPanel",
  props: {
    title: {
      type: String,
      default: null
    },
    description: {
      type: String,
      default: null
    },
    span: {
      type: Number,
      default: null,
      validator: isValidColSpan
    },
    condensed: {
      type: Boolean,
      default: false
    },
    condensedFooter: {
      type: Boolean,
      default: false
    }
  },
  computed: {
    bodyClasses() {
      const classes = !this.condensed ? [] : paddingClasses("none");
      return [`${PANEL_CLASS}__body`, ...classes];
    },
    footerClasses() {
      const classes = !this.condensedFooter ? [] : paddingClasses("none");
      return [`${PANEL_CLASS}__footer`, ...classes];
    },
    classes() {
      const staticClasses = [PANEL_CLASS];
      const { span } = this;
      return span == null
        ? staticClasses
        : [...staticClasses, `fd-has-grid-column-span-${span}`];
    },
    hasTitle() {
      return this.$slots.title != null || this.title != null;
    },
    hasDescription() {
      return this.$slots.description != null || this.description != null;
    },
    hasActions() {
      return this.$slots.actions != null;
    },
    needsHeader() {
      return this.needsHead || this.hasActions;
    },
    needsHead() {
      return this.hasTitle || this.hasDescription;
    }
  }
};
</script>
