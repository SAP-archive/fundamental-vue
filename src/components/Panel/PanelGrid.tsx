import { Component, DefaultSlot, Prop, Base } from '@/core';

const colMapping = {
  2: '2-Column Grid',
  3: '3-Column Grid',
  4: '4-Column Grid',
  5: '5-Column Grid',
  6: '6-Column Grid',
};
type Col = keyof typeof colMapping;
const Cols = Object.keys(colMapping).map(value => Number(value)) as Col[];

interface Props {
  col?: number | null;
  nogap?: boolean;
}

@Component('PanelGrid')
@DefaultSlot('Panels displayed by the grid.')
export class PanelGrid extends Base<Props> {
  @Prop('number of columns', {acceptableValues: Cols, type: Number, default: null })
  public col!: Col | null;

  @Prop('whether there is a gap between the individual panels', { type: Boolean, default: false })
  public nogap!: boolean;

  public render() {
    return <div class={this.classes}>{this.$slots.default}</div>;
  }

  private get classes() {
    const col = this.col;
    const colClass = col == null ? {} : { [`fd-panel-grid--${col}col`]: true };
    return {
      'fd-panel-grid': true,
      'fd-panel-grid--nogap': this.nogap,
      ...colClass,
    };
  }
}
