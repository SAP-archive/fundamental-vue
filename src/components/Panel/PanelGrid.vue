<template>
  <div :class="classes">
    <slot />
  </div>
</template>

<script>
const PANEL_GRID_CLASS = "fd-panel-grid";

export default {
  name: "FdPanelGrid",
  props: {
    col: {
      type: Number,
      default: null,
      validator: value => value >= 2 && value <= 6
    },
    nogap: { type: Boolean, default: false }
  },
  computed: {
    classes() {
      const col = this.col;
      return [
        PANEL_GRID_CLASS,
        ...(col != null ? [`${PANEL_GRID_CLASS}--${col}col`] : []),
        ...(this.nogap ? [`${PANEL_GRID_CLASS}--nogap`] : [])
      ];
    }
  }
};
</script>
