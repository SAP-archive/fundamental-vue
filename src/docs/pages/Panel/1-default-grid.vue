<title>Default Panel Grid (3 Columns)</title>

<template>
  <FdPanelGrid>
    <FdPanel>Panel Body 1</FdPanel>
    <FdPanel>Panel Body 2</FdPanel>
    <FdPanel>Panel Body 3</FdPanel>
    <FdPanel>Panel Body 4</FdPanel>
    <FdPanel>Panel Body 5</FdPanel>
  </FdPanelGrid>
</template>
