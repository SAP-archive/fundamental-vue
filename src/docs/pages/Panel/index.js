export default {
  relatedComponents: ["fd-panel", "fd-panel-grid"]
};
