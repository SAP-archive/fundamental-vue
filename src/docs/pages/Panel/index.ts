import { ExampleCollectionFunction } from '../types';

export const plugin: ExampleCollectionFunction = ({ Panel, PanelGrid }) => {
  return { icon: 'along-stacked-chart', relatedComponents: [Panel, PanelGrid] };
};
