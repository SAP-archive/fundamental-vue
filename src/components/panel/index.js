import Panel from './panel.vue'
import pluginify from './../../util/pluginify'
export default pluginify(Panel)
export { Panel }
