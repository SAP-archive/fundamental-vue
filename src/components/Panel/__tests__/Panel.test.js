import { mount } from "@vue/test-utils";
import Panel from "../Panel.vue";

describe("Panel", () => {
  const title = "Panel Title";
  const description = "Panel description";
  const actions = "Actions";
  const filters = "Filters";
  const body = "Body";
  const footer = "Footer";

  const wrapper = mount(Panel, {
    propsData: {
      title,
      description
    },
    slots: {
      actions: {
        render(h) {
          return h("p", {}, actions);
        }
      },
      filters: {
        render(h) {
          return h("p", {}, filters);
        }
      },
      default: {
        render(h) {
          return h("p", {}, body);
        }
      },
      footer: {
        render(h) {
          return h("p", {}, footer);
        }
      }
    }
  });

  it("renders correctly", () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it("renders correct props and slots", () => {
    expect(wrapper.find(".fd-panel__title").text()).toEqual(title);
    expect(wrapper.find(".fd-panel__description").text()).toEqual(description);
    expect(wrapper.find(".fd-panel__actions").text()).toEqual(actions);
    expect(wrapper.find(".fd-panel__filters").text()).toEqual(filters);
    expect(wrapper.find(".fd-panel__body").text()).toEqual(body);
    expect(wrapper.find(".fd-panel__footer").text()).toEqual(footer);
  });

  describe("Title as slot", () => {
    const title = "Panel Title (slot)";
    const wrapper = mount(Panel, {
      slots: {
        title: {
          render(h) {
            return h("div", {}, title);
          }
        }
      }
    });

    it("renders correctly", () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it("renders title as slot", () => {
      expect(wrapper.find(".fd-panel__title div").text()).toBe(title);
    });
  });

  describe("Description as slot", () => {
    const description = "Panel Description (slot)";
    const wrapper = mount(Panel, {
      slots: {
        description: {
          render(h) {
            return h("div", {}, description);
          }
        }
      }
    });

    it("renders correctly", () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it("renders title as slot", () => {
      expect(wrapper.find(".fd-panel__description div").text()).toBe(description);
    });
  });
});
