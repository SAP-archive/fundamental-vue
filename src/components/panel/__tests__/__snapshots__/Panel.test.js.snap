// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Panel Description as slot renders correctly 1`] = `
<div
  class="fd-panel"
>
  <div
    class="fd-panel__header"
  >
    <div
      class="fd-panel__head"
    >
      <!---->
       
      <p
        class="fd-panel__description"
      >
        <div>
          Panel Description (slot)
        </div>
      </p>
    </div>
     
    <!---->
  </div>
   
  <!---->
   
  <!---->
   
  <!---->
</div>
`;

exports[`Panel Title as slot renders correctly 1`] = `
<div
  class="fd-panel"
>
  <div
    class="fd-panel__header"
  >
    <div
      class="fd-panel__head"
    >
      <h1
        class="fd-panel__title"
      >
        <div>
          Panel Title (slot)
        </div>
      </h1>
       
      <!---->
    </div>
     
    <!---->
  </div>
   
  <!---->
   
  <!---->
   
  <!---->
</div>
`;

exports[`Panel renders correctly 1`] = `
<div
  class="fd-panel"
>
  <div
    class="fd-panel__header"
  >
    <div
      class="fd-panel__head"
    >
      <h1
        class="fd-panel__title"
      >
        
          Panel Title
        
      </h1>
       
      <p
        class="fd-panel__description"
      >
        
          Panel description
        
      </p>
    </div>
     
    <div
      class="fd-panel__actions"
    >
      <p>
        Actions
      </p>
    </div>
  </div>
   
  <div
    class="fd-panel__filters"
  >
    <p>
      Filters
    </p>
  </div>
   
  <div
    class="fd-panel__body"
  >
    <p>
      Body
    </p>
  </div>
   
  <div
    class="fd-panel__footer"
  >
    <p>
      Footer
    </p>
  </div>
</div>
`;
